src/run_autodiscovery.py

```diff
@@ -0,0 +1,237 @@
+import asyncio
+import json
+import os
+import re
+
+from dataclasses import dataclass
+from lib.clientsession_provider import init_gcp_client_session
+from lib.context import LoggingContext, MetricsContext, get_should_require_valid_certificate
+from lib.credentials import create_token, fetch_dynatrace_api_key, fetch_dynatrace_url
+from lib.metrics import Metric
+from typing import List, Text, Any, Dict
+from lib.configuration import config
+
+logging_context = LoggingContext(None)
+projectID="dynatrace-gcp-extension"
+discovered_resource_type="cloud_function"
+
+
+
+@dataclass(frozen=True)
+class MetricMetadata:
+    """Describes singular GCP service to ingest data from."""
+    # IMPORTANT! this object is only for one combination of object/featureSet!
+    # If you have 2 featureSets enabled for some service you will have 2 such objects
+
+    display_name: Text
+    description: Text
+    unit: Text
+    tags: List[Text]
+    dimensions: List[Dict[Text,Text]]
+    metric_properties: Dict[Text,Any]
+    schema_id: Text
+    scope: Text
+
+
+    def __init__(self, **kwargs):
+        object.__setattr__(self, "displayName", kwargs.get("displayName", ""))
+        object.__setattr__(self, "description", kwargs.get("description", ""))
+        object.__setattr__(self, "unit", kwargs.get("unit", "Unspecified"))
+        object.__setattr__(self, "tags", kwargs.get("tags", []))
+        object.__setattr__(self, "dimensions", kwargs.get("dimensions", []))
+        object.__setattr__(self, "metricProperties", kwargs.get("metricProperties", []))
+
+        object.__setattr__(self, "schemaId", kwargs.get("schemaId","builtin:metric.metadata"))
+        object.__setattr__(self,"scope","metric-"+kwargs.get("key", "").replace(":", "."))
+
+    def to_json(self):
+        return json.dumps({
+            "scope": self.scope,
+            "schema.Id": self.schema_id,
+            "value": {
+                "displayName": self.display_name,
+                "description": self.description,
+                "unit": self.unit,
+                "tags": self.tags,
+                "metricProperties": self.metric_properties,
+                "dimensions": self.dimensions
+            }
+        })
+
+
+def unit_caster(unit: str) -> str:
+    if unit in ["1","count","Count", "{operation}","{packet}" ,"{packets}" ,"{request}" ,"{port}" ,"{connection}", "{devices}" ,"{errors}" ,"{cpu}", "{query}", "{inode}"]:
```

Comment:
case ?

Reply:
I think it should be a dictionary, you look up by the key

Reply:
And also I would move all those dictionaries to separate file so it is easier to read the code